BUG: Fix NaT handling in the PyArray_CompareFunc for datetime and timedelta

asmeurer opened the pull request:
In #12658 and #15068 the sort ordering for datetime and timedelta was changed
so that NaT sorts to the end, but the internal compare array function was not
updated.

Fixes #19574.

If a test should be added for this, please let me know where. 

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

asmeurer committed:
BUG: Fix NaT handling in the PyArray_CompareFunc for datetime and timedelta

In #12658 and #15068 the sort ordering for datetime and timedelta was changed
so that NaT sorts to the end, but the internal compare array function was not
updated.

Fixes #19574.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2920,6 +2918,43 @@ C@TYPE@_compare(@type@ *pa, @type@ *pb)
+ * #type = npy_datetime, npy_timedelta#
+ */
+
+#define LT(a,b) ((a) < (b) || ((b) != (b) && (a) ==(a)))
```
I don't think you need this, the `((b) != (b) && (a) ==(a)))` part is for dealing with nans, which isn't a problem with integer types like `npy_datetime` and `npy_timedelta`.

asmeurer replied:
Ah you are right. I did notice that I copied the float template instead of the integer template.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2920,6 +2918,43 @@ C@TYPE@_compare(@type@ *pa, @type@ *pb)
+
+    printf("Calling @TYPE@_compare\n");
+    if (a == NPY_DATETIME_NAT) {
+        ret = 1;
```
What if both `a` and `b` are NaT? I note that sort order is not necessarily the same as comparison, which is one of the reasons only `LT`  is used for sorting.

asmeurer committed:
Remove debug print

asmeurer committed:
Fix the implementation of DATETIME_compare

I accidentally based it off the float compare template instead of the integer
compare template. It also now properly handles the case when both arguments
are NaT.

charris commented:
Thanks @asmeurer . This could probably be tested using an object array containing time* types, something like
```
In [20]: a = array([timedelta64('NaT'), timedelta64(0)], dtype=object)          

In [21]: a.sort() 
```
But that is hard to check without merging this.
